Eq 399 Refactoring Navigator component

warrenbailey opened the pull request:
### What is the context of this PR?

Refactoring the navigation and routing components and consolidating the user action functionality into a single place.

The navigation state object is now an actual state pattern, which maintains the current user position in the questionnaire.

The routing component was actually dealing with user actions which isn't its purpose. That should be handled elsewhere and it should only do routing (which we're yet to implement).

The questionnaire manager was also dealing with user actions making it difficult to isolate how they are processed.

The user actions are now handled by the navigator using a chain of responsibility pattern. This means adding a new user action should simply be a case of creating a new UserAction class and placing it in the correct location in the chain.

I appreciate this is a big refactor but I believe it leads us to a more elegant solution going forward. I'm open to comments as there's definitely still room for improvements.
### How to review
1. Check out this branch
2. Check all the tests still work
3. Complete a survey and check nothing is broken
- [ ] Do all commits have sensible commit messages?
- [ ] Is all new code unit tested?
- [ ] Are there integration tests if needed?
- [ ] Is there sufficient logging?
- [ ] Is there documentation or is the code self-describing?
### Who worked on the PR

@warrenbailey

warrenbailey committed:
Refactoring navigator to use a command pattern

warrenbailey committed:
Refactored Routing Engine to use a chain of responsibility

warrenbailey committed:
Refactoring Routing component into the navigator

warrenbailey committed:
Moving user actions to navigator, refactoring navigator to use a state pattern

warrenbailey commented:
Adding a WIP label but i'd still value some feedback on it, once I'm happy with it I'll squash the commits ready for merge.

weapdiv-david commented:
Looks good so far, although I don't fully understand it all yet.  It makes perfect sense to split the user action processing out from the data, and I'm currently looking at better ways to process the form data in view of the changes we want to make around fields/responses/questions.

Keep going!

warrenbailey commented:
Thanks for the feedback @weapdiv-david, if anyone else has time to review I'd appreciate it. Otherwise I'll fix the minor quantified code issues and squash the commits ready for merging.

warrenbailey commented:
Rebased and squashed the commits ready for merge. I liked to take this further but wonder if it would be better to merge this now and then make another incremental refactor, otherwise this risks getting out of date with master very quickly.

weapdiv-david commented:
Happy to merge this... Any other team members wanting to review it? (clue: you should)

dhilton commented on app/navigation/navigator.py:
```diff
@@ -1,82 +1,156 @@
-    def _valid_destination(self, destination):
-        """
-        Returns a boolean indicating whether a proposed destination is a valid one
+    def get_action(self):
```
Does this getter always return a string? If so, is it not a property that is set rather than a method?

warrenbailey replied:
Yep always returns a string. I'll change it to be a property.

dhilton commented:
Apart from my comment above 👍 from me.

warrenbailey committed:
Code review comments

warrenbailey commented:
Fixing code review comments and rebasing

dhilton commented:
Happy to merge this LGTM 👍
